### ISSUE TITLE
consult-{line,grep,ripgrep,...} and long lines with matches not showing

### ISSUE THREAD

**Issue by jdtsmith** [minad/consult#748]
If you use `consult-line` or `grep` etc. to search in an org or latex file with long lines (i.e. files you'd use with visual-line mode), it's strange that the text of some candidate lines have _no matches_ highlighted.  They of course do have matches, they're just off the window and inaccessible.

It would be helpful for such lines if consult `…left-truncated <match> lines so at least one of the <match>es shows up`.  Perhaps (as in this example) with a bit of context before the 1st match (say 25chars or 1/4 the window width, whichever is smaller).

**Comment by minad** [comment]
1. For `consult-line` truncation could only be implemented on the side of the UI. It is the same problem with the `consult-recent-file` left truncation. The candidates for `consult-line` are made up of all the full lines, which means that truncation must happen after matching.
2. For `consult-grep` we must truncate the lines even before they arrive in Emacs, since otherwise we seriously degrade performance. See `consult-grep-max-columns`.

In short, both issues are wontfix since long lines are not a common use case and Emacs has (or had) always problems with long lines.

**Comment by jdtsmith** [comment]
>  I wonder why your Org or LaTeX files have such long lines

This is on purpose. Long lines are very convenient in some situations.  I, like many, prefer using `visual-line-mode` where it makes sense, which means a "paragraph" is a line.  This is a common way of working in LaTeX.  One of the main advantages is that it is more friendly for version control: since paragraphs do not need to be "re-flowed", small changes show up as... small change diffs.   You also don't need to press M-q every few changes as you edit within paragraphs.  For org mode the main advantage is you can fit more text across the width of the buffer and never have "hanging fragments" when your window width is smaller than your line wrap.  This auto wrap as the window width changes works incredibly well in org mode, including for lists.

> truncation could only be implemented on the side of the UI

OK, I suppose in UI hooks I could look for the `consult-highlight-match` and do my truncation there.

**Comment by oantolin** [comment]
> One of the main advantages is that it is more friendly for version control: since paragraphs do not need to be "re-flowed", small changes show up as... small change diffs.

Even better for diffs: one line per sentence! I highly recommend one line per sentence though I can't bring myself to use that system (it just looks too weird)! 🤣

**Comment by minad** [comment]
Maybe we can come up with a generic solution for the truncation, which could be added to Vertico. One can use left truncation for file names and also if the matches occur too far to the right of the string. But maybe it will also look weird if the candidates move around.

**Comment by jdtsmith** [comment]
@oantolin Some of my colleagues actually do this; makes commenting out a sentence trivial (but they also write some _long_ sentences!).  Hard to look at indeed.  Once I adopted visual line mode for non-code, it just clicked.  

I tried my hand at left truncating via searching for highlighting, but got twisted in knots quickly.  Just too many different approaches among consult-line, consult-grep, etc.  If our general problem is "ensuring the first important match is visible to the user", perhaps a text property like `'ensure-column-visible' could be added by that matcher (e.g. the grep matcher).  UIs could then be opted-in to left-truncate with some padding to ensure that candidate column is visible.  They'll have to have a good model of what part of the candidate will be showing; not sure if that's the case right now. 

BTW, I still advise `vertico--format-candidate` with:

```elisp
  (defun my/vertico-truncate-candidates (args)
    (let* ((cand (car args))
	   (l (length cand))
	   (ww (window-width))
	   type)
      (cond
       ;; left-truncate recent files in consult-buffer
       ((and (setq type (get-text-property 0 'multi-category cand))
	     (eq (car-safe type) 'file))
	(let ((w (max 30 (- ww (consult--display-width (nth 2 args))))))
	  (when (> l w)
	    (setcar args 
		    (concat "…" (truncate-string-to-width cand l (- l w)))))))
...
```

**Comment by minad** [comment]
For consult-grep it would be possible but as I mentioned long candidates don't even arrive in Emacs. For consult-line we cannot do anything, since there the marker would have to be added by Orderless and all the other completion styles. I am afraid we cannot solve this easily. My recommendation is to either use line breaks or to use Isearch.

Regarding diffing did you try the git diff option which ignores whitespace?

**Comment by minad** [comment]
Another more realistic option could be to not truncate lines in Vertico, which could be enabled via vertico-multiform. It would require changes to the resizing logic though. But it would probably hurt the completion UI too much anyway.

**Comment by oantolin** [comment]
I hate to say it, but these long candidates doing like a job for the defunct embark-collect-completions with zebra mode.

**Comment by minad** [comment]
@oantolin Indeed. I am still not happy with the current state of the zebra and would prefer an improved meerkat mode. Overlays are not ideal for this and it would be nice to have an option to use separator lines. Maybe one could even automatically enable the mode the candidates contain newlines or wrap around?

**Comment by jdtsmith** [comment]
I distinguish between "foolishly, unnecessarily long" lines of the sort that tracebacks produce (thousands or 10s of thousands of chars) and "long lines of artful intent".  For example, in a long org file of mine, my median, mean, max line lengths are 28, 65.72, 709, with 15.6% of the lines >120 chars.  So all of these would be delivered by grep.  As nice as consult-line/consult-grep etc. are, the value of visual-line-mode is just too great to give up.

What about (opting-in) to use visual-line mode in the minibuffer, if candidate lines exceed some maximum length?  Consult already trims lines to 1000chars by default.  `forward-line` easily skips between such 'wrapped lines".  That said, a strategy to left-truncates to ensure the first match is visible is probably 99% of the way there for me; preview after all can show you exactly how things look in situ.  And/or exporting to an embark mode that has visual-line-mode on (hadn't heard of the zebra...).

**Comment by minad** [comment]
@jdtsmith If you come up with some reasonable patches I am happy to consider them. But I don't see an easy solution. For me personally the problem is not that relevant since I usually use wrapped lines.

**Comment by jdtsmith** [comment]
I spent some time on it last night, and I'm afraid it will be hard for me to find a general solution, given the variety of different chefs with their spoons in the pot across all the different "select among lines" tools.  I'll keep it in the back of my mind.

**Comment by minad** [comment]
Yes, as I said, there is no easy solution. Maybe we just have to accept that we are hitting the limitations of the systems. Overall I agree with the assessment that there are long lines and foolishly long lines. See the value of consult-grep-max-columns, which is the limit for the fools. You could very well increase it to ~1000 and consult-grep should still be sufficiently fast. But the main priority of completion UIs is still to handle candidates of length < 100 well, candidates which fit into the window. Everything above is a bonus and not the main focus. Note that long candidate strings can also slow down filtering considerably, since Orderless matching speed slows down proportionally to the candidate length.

**Comment by jdtsmith** [comment]
Turns out the "check for changes the match highlighter made" approach actually works well for all completion styles.  I threw something together into a [small mode](https://github.com/jdtsmith/vertico-truncate) which patches `vertico--recompute` (to wrap the highlighter) and `vertico--format-candidate` (to do the truncation if needed).

The only wrinkle I found: `consult-grep` and friends do not defer grep-match-highlighting to the UI (candidates come pre-highlit), and also do not use a prefix, leaving grep's line numbers in the candidate.  Probably performance motivations for both.  Relatively easy to workaround.

Happy for this to be used however.

**Comment by minad** [comment]
@jdtsmith It is actually not that bad to use the vertico--highlight function to determine the location of the match. But overall it is quite ad-hoc. Maybe one can generalize everything a little bit.

1. For recent files we use left truncation. For some categories one would want left truncation for some others right truncation. Either one introduces a category alist or a variable, which is then set via vertico-multiform. 
2. If a long candidate is truncated we can move it around such that the first completion style highlight becomes visible.
3. For grep one needs an additional ad-hoc solution, e.g., one could introduce a text property which marks the important part of the candidate or one could use a variable as you have it, which lists the important faces.
4. Wrapping the highlight function is not that bad, but it is still feels like a hack. Note that I use cl-defmethod to override for example the display function in extensions (no advices anymore). This is a relatively clean approach I think.

Not sure if we should turn this into a Vertico extension, since I am not too fond of adding functionality to Vertico which is mostly (or only) useful for Consult commands. The feature (1) to truncate candidates alone could possibly go into Vertico as is (a configuration variable to define the maximum candidate length and left/right truncation), but the feature to move to the match feels more Consult-specific. Usually knowing the location of a match is not that of an interesting information for normal completion, while for consult-line with your use case this information matters more since you want to see the match and the context. It is a difference between filtering and searching.

**Comment by jdtsmith** [comment]
Some generalization and hooks would be a good idea.

> If a long candidate is truncated we can move it around such that the first completion style highlight becomes visible.

Exactly, that's (pretty much all) this does.  It does require some care to identify the "first match".  I had assumed all highlighting would be deferred to a final `vertico--highlight`, but for grep the "1st stage" is pre-matched and highlighted.   So you have to search for these hard-coded extra 1st-round highlighting faces (well, one so far).

> For grep one needs an additional ad-hoc solution, e.g., one could introduce a text property which marks the important part of the candidate or one could use a variable as you have it, which lists the important faces.

Ideally we could settle on a single property for this, so no faces have to be listed or maintained.  Then _any_ future modes from the consult/etc. universe could opt in by "marking the important part" they'd like shifted into view, if possible.   Maybe even orderless could do so ;).

Of course that leaves completion-styles, but it was a good puzzle to work out how to monitor their effects; works quite well (but beware: `cl-loop for var being the intervals` neglects to let-bind `var`!  Report submitted.).

> Wrapping the highlight function is not that bad, but it is still feels like a hack. 

Yeah, totally.  I kept having the feeling: this information should be knowable with spying on the highlighter.  But so many different pieces of code can add highlighting, some deep in emacs internals.  So it seemed the only foolproof way, without hard-coding a bunch of yucky faces that could change.  

It should be easy to drop the advice with a little vertico help.  One of the advantages of your abiding love for lambdas (which was the subject of a recent reddit post, btw!).  

> Not sure if we should turn this into a Vertico extension; <it's> mostly (or only) useful for Consult commands

I had this notion as well.  But on the other hand it sits purely in the realm of UI display, so it makes little sense to put it in consult.  If some fancier UI eclipses Vertico some day, it will presumably want "the whole candidate, please" for its three-dimensional hyperspectral interface ;).

**Comment by jdtsmith** [comment]
One issue I discovered: when using vertico-multiform with `buffer`, the reported `(window-width)` in the minibuffer-in-a-buffer is incorrectly returning the full frame width.  If you have left + right window layout, this is quite different (e.g. 80 vs. 180 chars). Perhaps this is because it _starts_ in the minibuffer, but is transferred to a buffer?  How can I get the "real" window width of the minibuffer displayed in a buffer?

**Comment by minad** [comment]
@jdtsmith See `vertico--window-width`.

**Comment by minad** [comment]
Please add it to the wiki for now. We can maybe add truncation to Vertico itself in the future, if we figure out a solution, which is less ad-hoc.

**Comment by jdtsmith** [comment]
Added

**Comment by dsedivec** [comment]
Hi from over a year later!  Here's why this problem is particularly bad, from my perspective:

1.  You run `consult-ripgrep` to grep through files with long lines
2. You use Embark to export the results to a buffer
3. You use wgrep to edit those lines in the export
4. wgrep fails to edit your lines because some of them were truncated, and so wgrep can't find the line it thinks you were modifying (with a message like "There are 27 unapplied changes. (0 changed)")

Since the export buffer opened, and wgrep started, I just *assumed* that I would be able to edit all these lines in all these files.  My bad.

Is there any way to have `consult-grep` and friends export the complete long line?

**Comment by minad** [comment]
@dsedivec Disabling truncation may help: https://github.com/minad/consult/commit/0557ff36c0e868c92ebac078e7c3ef3bad61f055
